src/python/zquantum/core/decompositions/_zquantum_decompositions.py

```diff
@@ -0,0 +1,47 @@
+from typing import Iterable
+
+from zquantum.core.circuits._builtin_gates import RY, RZ
+from zquantum.core.circuits._circuit import Circuit
+from zquantum.core.circuits._gates import GateOperation
+from zquantum.core.decompositions._decomposition import (
+    DecompositionRule,
+    decompose_operations,
+)
+
+
+class U3GateToRotation(DecompositionRule[GateOperation]):
+    """Decomposition of ZQuantum's U3 gate.
+
+    Note that this gets rid of global phase.
+    """
+
+    def predicate(self, operation: GateOperation) -> bool:
+        # Only decompose U3 and its controlled version
+        return (
+            operation.gate.name == "U3"
+            or operation.gate.name == "Control"
+            and operation.gate.wrapped_gate.name == "U3"
+        )
+
+    def production(self, operation: GateOperation) -> Iterable[GateOperation]:
+        theta, phi, lambda_ = operation.params
+
+        decomposition = [RZ(phi), RY(theta), RZ(lambda_)]
+
+        preprocessing = (
+            lambda x: x.controlled(operation.gate.num_control_qubits)
+            if operation.gate.name == "Control"
+            else x
```

Comment:
In most of our codebase, we don't use lambdas. Instead, we typically use locally defined functions. WDYT about doing it here?